Fix option admin names collisions due to case insensitivity during CSV export

**Describe the bug**
Anytime a new Admin name is created for an multiselect question, we should not only check for its existence but also include case sensitivity check with existing ones.

**To Reproduce**
Steps to reproduce the behavior:

1. Create a multiselect question with two or three options with admin names like Red_color and Green_color
2. Add another option with admin name like red_color (case is different)
3. Create a program with this question and publish. Login as an applicant and sumit an application to this program.
4. Login as Program Admin and hit CSV export. you will see a timeout issue due to the export keys being the same.

**Expected behavior**
The second option red_color should not be permitted by the system as it has the same admin name as the previous option. 

**Screenshots**
![image](https://github.com/civiform/civiform/assets/101214724/b76ed46e-b53d-450b-925b-9299ed3ff1f9)

Error message - 
```
Caused by: java.lang.IllegalArgumentException: Multiple entries with same key: applicant.enumhousehold_member[1].interest_in_programs.recreation & aquatics programs (all ages)=SELECTED and applicant.enumhousehold_member[1].interest_in_programs.recreation & aquatics programs (all ages)=NOT_AN_OPTION_AT_PROGRAM_VERSION
	at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:378)
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:372)
	at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:246)
	at com.google.common.collect.RegularImmutableMap.fromEntryArrayCheckingBucketOverflow(RegularImmutableMap.java:133)
	at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:95)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:572)
	at com.google.common.collect.ImmutableMap$Builder.buildOrThrow(ImmutableMap.java:600)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:587)
	at services.export.CsvExporter.exportRecord(CsvExporter.java:118)
	at services.export.CsvExporterService.exportCsv(CsvExporterService.java:193)
	at services.export.CsvExporterService.getProgramAllVersionsCsv(CsvExporterService.java:108)
	at controllers.admin.AdminApplicationController.downloadAll(AdminApplicationController.java:187)
	at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$229(Routes.scala:4080)
	at play.core.routing.HandlerInvokerFactory$$anon$8.resultCall(HandlerInvoker.scala:150)
	at play.core.routing.HandlerInvokerFactory$$anon$8.resultCall(HandlerInvoker.scala:149)
	at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115)
	at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
	at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:31)
	at org.pac4j.play.java.SecureAction.lambda$internalCall$1(SecureAction.java:100)
	at org.pac4j.core.engine.DefaultSecurityLogic.perform(DefaultSecurityLogic.java:141)
```

--- a/server/app/services/question/types/QuestionDefinition.java
+++ b/server/app/services/question/types/QuestionDefinition.java
@@ -291,20 +291,20 @@ public abstract class QuestionDefinition {
 
       var existingAdminNames =
           previousDefinition
-              .map(qd -> (MultiOptionQuestionDefinition) qd)
               .map(
-                  result -> {
+                  qd -> {
                     ImmutableList<String> optionList =
-                        multiOptionQuestionDefinition.getOptionAdminNames();
+                        ((MultiOptionQuestionDefinition) qd).getOptionAdminNames();
                     return optionList.stream()
                         .map(e -> e.toLowerCase(Locale.ROOT))
                         .collect(Collectors.toUnmodifiableList());
                   })
               .orElse(ImmutableList.of());
+
       if (multiOptionQuestionDefinition.getOptionAdminNames().stream()
           // This is O(n^2) but the list is small and it's simpler than creating a Set
           .filter(n -> !existingAdminNames.contains(n.toLowerCase(Locale.ROOT)))
-          .anyMatch(s -> !s.matches("[0-9a-z_-]+"))) {
+          .anyMatch(s -> !s.matches("[0-9a-zA-Z_-]+"))) {
         errors.add(
             CiviFormError.of(
                 "Multi-option admin names can only contain letters, numbers, underscores, and"
@@ -319,7 +319,7 @@ public abstract class QuestionDefinition {
       long numUniqueOptionDefaultValues =
           multiOptionQuestionDefinition.getOptions().stream()
               .map(QuestionOption::optionText)
-              .map(LocalizedStrings::getDefault)
+              .map(e -> e.getDefault().toLowerCase(Locale.ROOT))
               .distinct()
               .count();
       if (numUniqueOptionDefaultValues != numOptions) {
@@ -329,6 +329,7 @@ public abstract class QuestionDefinition {
       long numUniqueOptionAdminNames =
           multiOptionQuestionDefinition.getOptions().stream()
               .map(QuestionOption::adminName)
+              .map(e -> e.toLowerCase(Locale.ROOT))
               .distinct()
               .count();
       if (numUniqueOptionAdminNames != numOptions) {

--- a/server/app/services/question/types/QuestionDefinition.java
+++ b/server/app/services/question/types/QuestionDefinition.java
@@ -304,7 +304,7 @@ public abstract class QuestionDefinition {
       if (multiOptionQuestionDefinition.getOptionAdminNames().stream()
           // This is O(n^2) but the list is small and it's simpler than creating a Set
           .filter(n -> !existingAdminNames.contains(n.toLowerCase(Locale.ROOT)))
-          .anyMatch(s -> !s.matches("[0-9a-zA-Z_-]+"))) {
+          .anyMatch(s -> !s.matches("[0-9a-z_-]+"))) {
         errors.add(
             CiviFormError.of(
                 "Multi-option admin names can only contain letters, numbers, underscores, and"

--- a/server/app/services/question/types/QuestionDefinition.java
+++ b/server/app/services/question/types/QuestionDefinition.java
@@ -15,6 +15,7 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.Random;
+import java.util.stream.Collectors;
 import services.CiviFormError;
 import services.LocalizedStrings;
 import services.Path;
@@ -291,18 +292,24 @@ public abstract class QuestionDefinition {
       var existingAdminNames =
           previousDefinition
               .map(qd -> (MultiOptionQuestionDefinition) qd)
-              .map(MultiOptionQuestionDefinition::getOptionAdminNames)
+              .map(
+                  result -> {
+                    ImmutableList<String> optionList =
+                        multiOptionQuestionDefinition.getOptionAdminNames();
+                    return optionList.stream()
+                        .map(e -> e.toLowerCase(Locale.ROOT))
+                        .collect(Collectors.toUnmodifiableList());
+                  })
               .orElse(ImmutableList.of());
       if (multiOptionQuestionDefinition.getOptionAdminNames().stream()
           // This is O(n^2) but the list is small and it's simpler than creating a Set
-          .filter(n -> !existingAdminNames.contains(n))
+          .filter(n -> !existingAdminNames.contains(n.toLowerCase(Locale.ROOT)))
           .anyMatch(s -> !s.matches("[0-9a-zA-Z_-]+"))) {
         errors.add(
             CiviFormError.of(
                 "Multi-option admin names can only contain letters, numbers, underscores, and"
                     + " dashes"));
       }
-
       if (multiOptionQuestionDefinition.getOptions().stream()
           .anyMatch(option -> option.optionText().hasEmptyTranslation())) {
         errors.add(CiviFormError.of("Multi-option questions cannot have blank options"));

--- a/server/test/services/question/types/QuestionDefinitionTest.java
+++ b/server/test/services/question/types/QuestionDefinitionTest.java
@@ -444,6 +444,22 @@ public class QuestionDefinitionTest {
   }
 
   @Test
+  public void validate_multiOptionQuestion_withDuplicateOptionsWithDifferentCase_returnsError() {
+    QuestionDefinitionConfig config = configBuilder.build();
+    ImmutableList<QuestionOption> questionOptions =
+        ImmutableList.of(
+            QuestionOption.create(
+                1L, "Parks_and_Recreation", LocalizedStrings.withDefaultValue("a")),
+            QuestionOption.create(
+                2L, "Parks_and_recreation", LocalizedStrings.withDefaultValue("a")));
+    QuestionDefinition question =
+        new MultiOptionQuestionDefinition(
+            config, questionOptions, MultiOptionQuestionType.CHECKBOX);
+    assertThat(question.validate())
+        .containsOnly(CiviFormError.of("Multi-option question options must be unique"));
+  }
+
+  @Test
   public void validate_multiOptionQuestion_withDuplicateOptionAdminNames_returnsError() {
     QuestionDefinitionConfig config = configBuilder.build();
     ImmutableList<QuestionOption> questionOptions =
@@ -525,6 +541,37 @@ public class QuestionDefinitionTest {
 
   @Test
   public void
+      validate_multiOptionQuestion_withValidOptionInPreviousDraftSameOptionWithDifferentCaseInNewOption_ReturnsError() {
+    QuestionDefinitionConfig config =
+        QuestionDefinitionConfig.builder()
+            .setName("test")
+            .setDescription("test")
+            .setQuestionText(LocalizedStrings.withDefaultValue("test"))
+            .setQuestionHelpText(LocalizedStrings.empty())
+            .build();
+    ImmutableList<QuestionOption> previousQuestionOptions =
+        ImmutableList.of(
+            QuestionOption.create(1L, "OptionA", LocalizedStrings.withDefaultValue("a")),
+            QuestionOption.create(2L, "OptionB", LocalizedStrings.withDefaultValue("b")));
+    QuestionDefinition previousQuestion =
+        new MultiOptionQuestionDefinition(
+            config, previousQuestionOptions, MultiOptionQuestionType.CHECKBOX);
+
+    ImmutableList<QuestionOption> updatedQuestionOptions =
+        ImmutableList.<QuestionOption>builder()
+            .addAll(previousQuestionOptions)
+            .add(QuestionOption.create(2L, "optionA", LocalizedStrings.withDefaultValue("optiona")))
+            .build();
+    QuestionDefinition updatedQuestion =
+        new MultiOptionQuestionDefinition(
+            config, updatedQuestionOptions, MultiOptionQuestionType.CHECKBOX);
+
+    assertThat(updatedQuestion.validate(Optional.of(previousQuestion)))
+        .containsOnly(CiviFormError.of("Multi-option question admin names must be unique"));
+  }
+
+  @Test
+  public void
       validate_multiOptionQuestion_withInvalidOptionAdminNameInPreviousAndUpdatedDefinition_returnsError() {
     QuestionDefinitionConfig config =
         QuestionDefinitionConfig.builder()
@@ -549,7 +596,6 @@ public class QuestionDefinitionTest {
     QuestionDefinition updatedQuestion =
         new MultiOptionQuestionDefinition(
             config, updatedQuestionOptions, MultiOptionQuestionType.CHECKBOX);
-
     assertThat(updatedQuestion.validate(Optional.of(previousQuestion)))
         .containsOnly(
             CiviFormError.of(

--- a/server/test/services/question/types/QuestionDefinitionTest.java
+++ b/server/test/services/question/types/QuestionDefinitionTest.java
@@ -449,14 +449,14 @@ public class QuestionDefinitionTest {
     ImmutableList<QuestionOption> questionOptions =
         ImmutableList.of(
             QuestionOption.create(
-                1L, "Parks_and_Recreation", LocalizedStrings.withDefaultValue("a")),
+                1L, "Parks_and_Recreation", LocalizedStrings.withDefaultValue("a1")),
             QuestionOption.create(
-                2L, "Parks_and_recreation", LocalizedStrings.withDefaultValue("a")));
+                2L, "Parks_and_recreation", LocalizedStrings.withDefaultValue("a2")));
     QuestionDefinition question =
         new MultiOptionQuestionDefinition(
             config, questionOptions, MultiOptionQuestionType.CHECKBOX);
     assertThat(question.validate())
-        .containsOnly(CiviFormError.of("Multi-option question options must be unique"));
+        .contains(CiviFormError.of("Multi-option question admin names must be unique"));
   }
 
   @Test
@@ -541,7 +541,7 @@ public class QuestionDefinitionTest {
 
   @Test
   public void
-      validate_multiOptionQuestion_withValidOptionInPreviousDraftSameOptionWithDifferentCaseInNewOption_ReturnsError() {
+      validate_multiOptionQuestion_withValidOptionAdminNameInPreviousAndDuplicateNameInUpdate_returnsError() {
     QuestionDefinitionConfig config =
         QuestionDefinitionConfig.builder()
             .setName("test")
@@ -551,8 +551,8 @@ public class QuestionDefinitionTest {
             .build();
     ImmutableList<QuestionOption> previousQuestionOptions =
         ImmutableList.of(
-            QuestionOption.create(1L, "OptionA", LocalizedStrings.withDefaultValue("a")),
-            QuestionOption.create(2L, "OptionB", LocalizedStrings.withDefaultValue("b")));
+            QuestionOption.create(1L, "a_valid", LocalizedStrings.withDefaultValue("a")),
+            QuestionOption.create(2L, "b_valid", LocalizedStrings.withDefaultValue("b")));
     QuestionDefinition previousQuestion =
         new MultiOptionQuestionDefinition(
             config, previousQuestionOptions, MultiOptionQuestionType.CHECKBOX);
@@ -560,7 +560,7 @@ public class QuestionDefinitionTest {
     ImmutableList<QuestionOption> updatedQuestionOptions =
         ImmutableList.<QuestionOption>builder()
             .addAll(previousQuestionOptions)
-            .add(QuestionOption.create(2L, "optionA", LocalizedStrings.withDefaultValue("optiona")))
+            .add(QuestionOption.create(2L, "A_valid", LocalizedStrings.withDefaultValue("c")))
             .build();
     QuestionDefinition updatedQuestion =
         new MultiOptionQuestionDefinition(
@@ -596,6 +596,7 @@ public class QuestionDefinitionTest {
     QuestionDefinition updatedQuestion =
         new MultiOptionQuestionDefinition(
             config, updatedQuestionOptions, MultiOptionQuestionType.CHECKBOX);
+
     assertThat(updatedQuestion.validate(Optional.of(previousQuestion)))
         .containsOnly(
             CiviFormError.of(

